rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1540,6 +1558,258 @@ RMW_Connext_Subscriber::take_serialized(
+    DDS_Subscriber_delete_datareader(this->dds_subscriber(), this->dds_reader))
+  {
+    RMW_CONNEXT_LOG_ERROR_SET(
+      "failed to delete DDS DataReader")
+    return RMW_RET_ERROR;
+  }
+
+  DDS_DomainParticipant * const dp = this->dds_participant();
+  DDS_TopicDescription * sub_topic = nullptr;
+  if (nullptr == this->dds_topic_cft) {
+    // case 3: create a new cft if there is no cft topic and filter_expression is not empty("")
+    // remove old reader
+    // create cft topic and a new reader
+    std::string cft_topic_name =
+      fqtopic_name + "_ContentFilterTopic" + RMW_Connext_Subscriber::get_atomic_id();
+
+    rmw_ret_t cft_rc =
+      rmw_connextdds_create_contentfilteredtopic(
+      ctx, dp, dds_topic, cft_topic_name.c_str(),
+      filter_expression,
+      expression_parameters, &sub_topic);
+
+    if (RMW_RET_OK != cft_rc) {
+      RMW_CONNEXT_LOG_ERROR("failed to create content filter topic")
+      return RMW_RET_ERROR;
+    }
+
+    this->dds_topic_cft = sub_topic;
+  } else {
+    // case 4: delete cft topic and create a normal reader if filter_expression is empty("")
+    // that means reset
+    // remove content filter topic
+    if (nullptr != this->dds_topic_cft) {
+      rmw_ret_t cft_rc = rmw_connextdds_delete_contentfilteredtopic(
+        ctx, dp, this->dds_topic_cft);
+
+      if (RMW_RET_OK != cft_rc) {
+        return cft_rc;
+      }
+    }
+
+    this->dds_topic_cft = nullptr;
+
+    // create normal topic and a new reader
+    sub_topic = DDS_Topic_as_topicdescription(dds_topic);
+  }
+
+#if !RMW_CONNEXT_DDS_API_PRO_LEGACY
+  DDS_DataReaderQos dr_qos = DDS_DataReaderQos_INITIALIZER;
+#else
+  DDS_DataReaderQos dr_qos;
+  if (DDS_RETCODE_OK != DDS_DataReaderQos_initialize(&dr_qos)) {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to initialize datareader qos")
+    return RMW_RET_ERROR;
+  }
+#endif /* !RMW_CONNEXT_DDS_API_PRO_LEGACY */
+
+  DDS_DataReaderQos * const dr_qos_ptr = &dr_qos;
+  auto scope_exit_dr_qos_delete =
+    rcpputils::make_scope_exit(
+    [dr_qos_ptr]()
+    {
+      DDS_DataReaderQos_finalize(dr_qos_ptr);
+    });
+
+  if (DDS_RETCODE_OK !=
+    DDS_Subscriber_get_default_datareader_qos(this->dds_subscriber(), &dr_qos))
+  {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to get default reader QoS")
+    return RMW_RET_ERROR;
+  }
+
+  DDS_DataReader * dds_reader =
+    rmw_connextdds_create_datareader(
+    ctx,
+    dp,
+    this->dds_subscriber(),
+    &qos_policies,
+#if RMW_CONNEXT_HAVE_OPTIONS_PUBSUB
+    nullptr,
+#endif /* RMW_CONNEXT_HAVE_OPTIONS_PUBSUB */
+    internal,
+    type_support,
+    sub_topic,
+    &dr_qos);
+
+  if (nullptr == dds_reader) {
+    RMW_CONNEXT_LOG_ERROR_SET("failed to create DDS reader")
+    return RMW_RET_ERROR;
+  }
+
+  this->dds_reader = dds_reader;
+
+  this->status_condition.reset(
+    new RMW_Connext_SubscriberStatusCondition(this->dds_reader, this->ignore_local));
+  rmw_connextdds_get_entity_gid(this->dds_reader, this->ros_gid);
+
+  if (!internal) {
```

Comment:
In general, I wouldn't expect this function to ever be called on "internal" subscribers (which, at the moment, is only one DataReader created by the RMW for topic `ros_discovery_info`). This check can be removed, and I would instead add a "precondition check" at the beginning of the function to return with an error if the function is called on an "internal" endpoint (it could maybe be an assertion with `RMW_CONNEXT_ASSERT(!internal)`, since there is no way to call it on these endpoints)

Reply:
good to know, thanks.